cpp/src/file_handle.cpp

```diff
@@ -118,26 +119,28 @@ int open_fd(const std::string& file_path, const std::string& flags, bool o_direc
 FileHandle::FileHandle(const std::string& file_path,
                        const std::string& flags,
                        mode_t mode,
-                       bool compat_mode)
+                       CompatMode compat_mode)
   : _fd_direct_off{open_fd(file_path, flags, false, mode)},
     _initialized{true},
     _compat_mode{compat_mode}
 {
-  if (_compat_mode) {
+  if (_compat_mode == CompatMode::ON) {
     return;  // Nothing to do in compatibility mode
+  } else if (_compat_mode == CompatMode::AUTO) {
+    _compat_mode = defaults::infer_compat_mode_from_runtime_sys(_compat_mode);
```

Comment:
why not infer in `_compat_mode` init?

Reply:
I've brought the infer to the beginning of the constructor. Is this what you would recommend?

Reply:
I meant the line `_compat_mode{compat_mode}`, but it looks like we don't want to infer at all to ensure correct fallback.